### DISCUSSION TITLE
Making a Rewinder Gradient

### DISCUSSION THREAD

**Discussion by nathanjbeaumont** [pulseq/pulseq#63] [Q&A] [answered]
Hello,
I have a spiral sequence which can end at any point in k-space I need to make a rewinder gradient to get back to the center of k-space given a max slew rate and max gradient.

I found the function `makeExtendedTrapezoidArea` which could do what if I set `Gs` to my spiral's final Gx, `Ge` to 0, and `A` to the negative spiral area. However the function constantly throws up errors with its asserts. Is there another function I could use for rewinding?

**Comment by QingpingChen1201** [comment]
Hi,
the way how you use the function `makeExtendedTrapezoidArea `sounds correct. This function is the right function for generating a rewinder starting from non-zero with a specific area. Could you please screenshot the errors?
Best regards,
Qingping

**Comment by m-a-x-i-m-z** [comment]
Hi, 

Frank Zijlstra has told me that the PyPulseq version of this function had many problems in the past, which is why about half a year ago he had done quite a bit of debugging and basically rewritten the function (but only in PyPulseq, not in Matlab). His point was that although the pulse parameters are derived from a solution of a relatively simple quadratic equation, the requirement of aligning the time to the gradient raster makes the problem highly non-convex and that for some combinations of the start and end gradient strengths and required gradient moment only grid search is able to find a solution. 

Interestingly, we haven't observed that many problems with the Matlab's implementation, so we didn't rush to convert the Python code to Matlab. 

So, one option, obliviously, would be  to reprogram the current algorithm in the Matlab function based on the PyPulseq counterpart. 

Another option, if you don't insist on having the optimal timing is to ramp down the gradient to zero in the shortest possible time and then apply a simple trapezoid to generate the residual required moment. 

And, as Qungping says, it would be interesting to have a couple of test cases to see which parameters cause the current function to fail. From our discussions with Frank my understanding is that it is particularly difficult for the optimization algorithm to find a solution that for example stars from a positive gradient amplitude and has to generate a ramp-down with a small negative blip. But, maybe there are further challenging cases.

**Comment by QingpingChen1201** [comment, marked as answer]
Hi all,
Thanks Maxim for the inforamtion. By coincidence I also encountered some converge errors when I used the makeExtendedTrapezoidArea function today. And then I used the script updated by Mehmet Emin Öztürk, who is a Master student from Bilkent University, Turkey. The updated function works pretty well. I attach the updated script he sent me here in case you would need it quite urgently. Later we can merge the function carefully and update it in the Github repository.
[makeExtendedTrapezoidArea_v2.zip](https://github.com/user-attachments/files/17534417/makeExtendedTrapezoidArea_v2.zip)
Best regards,
Qingping

**Reply by nathanjbeaumont** [reply]
Using this function is working perfectly without issues. Thanks for this.
